File: day13.lua
#!/usr/bin/env lua

local function solve(game)
    local a, c = table.unpack(game[1])
    local b, d = table.unpack(game[2])
    local x, y = table.unpack(game[3])
    local z = d * x - b * y
    local w = a * y - c * x
    local det = a * d - b * c
    if z % det ~= 0 or w % det ~= 0 then return 0 end
    return (z * 3 + w) // det
end

local file = io.open("day13.txt")
local sum = { 0, 0 }

repeat
    local game = {
        { file:read():match("Button A: X%+(%d+), Y%+(%d+)") },
        { file:read():match("Button B: X%+(%d+), Y%+(%d+)") },
        { file:read():match("Prize: X=(%d+), Y=(%d+)") },
    }
    for i = 1, #game do
        for j = 1, #game[i] do
            game[i][j] = tonumber(game[i][j])
        end
    end

    sum[1] = sum[1] + solve(game)
    game[3][1] = game[3][1] + 10000000000000
    game[3][2] = game[3][2] + 10000000000000
    sum[2] = sum[2] + solve(game)
until not file:read()

print(sum[1], sum[2])

File: day13.c
#include <stdio.h>

long long solve(long long (*game)[2])
{
    long long a = game[0][0], b = game[1][0], x = game[2][0];
    long long c = game[0][1], d = game[1][1], y = game[2][1];
    long long z = d * x - b * y;
    long long w = a * y - c * x;
    long long det = a * d - b * c;
    if (z % det || w % det) return 0;
    return (z * 3 + w) / det;
}

int main()
{
    FILE* file = fopen("day13.txt", "r");
    long long sum = 0, sum2 = 0;
    long long game[3][2];

    while (2 == fscanf(file, "Button A: X+%lld, Y+%lld\n", &game[0][0], &game[0][1])) {
        fscanf(file, "Button B: X+%lld, Y+%lld\n", &game[1][0], &game[1][1]);
        fscanf(file, "Prize: X=%lld, Y=%lld\n", &game[2][0], &game[2][1]);

        //for (int i = 0; i < 3; i++)
        //for (int j = 0; j < 2; j++)
            //printf("%lld ", game[i][j]);
        //printf("\n");

        sum += solve(game);
        game[2][0] += 10000000000000;
        game[2][1] += 10000000000000;
        sum2 += solve(game);
    }

    printf("%lld %lld", sum, sum2);
}
